SPARK-5143 [BUILD] [WIP] spark-network-yarn 2.11 depends on spark-network-shuffle 2.10

srowen opened the pull request:
Update `<scala.binary.version>` prop in POM when switching between Scala 2.10/2.11

@ScrapCodes for review. This `sed` command is supposed to just replace the first occurrence, but it replaces them all. Are you more of a `sed` wizard than I? It may be a GNU/BSD thing that is throwing me off. Really, just the first instance should be replaced, hence the `[WIP]`.

NB on OS X the original `sed` command here will create files like `pom.xml-e` through the source tree though it otherwise works. It's like `-e` is also the arg to `-i`. I couldn't get rid of that even with `-i""`. No biggie.

srowen committed:
Update scala.binary.version prop in POM when switching between Scala 2.10/2.11

SparkQA commented:
[Test build #28234 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28234/consoleFull) for   PR 4876 at commit [`703e1eb`](https://github.com/apache/spark/commit/703e1ebf7fde1044a800ad07070fa57704f98fad).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28234 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28234/consoleFull) for   PR 4876 at commit [`703e1eb`](https://github.com/apache/spark/commit/703e1ebf7fde1044a800ad07070fa57704f98fad).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28234/
Test PASSed.

ScrapCodes commented on dev/change-version-to-2.10.sh:
```diff
@@ -17,4 +17,7 @@
+  | xargs -I {} sed -i -e 's|\(artifactId.*\)_2.11|\1_2.10|g' {}
+
+# Also update <scala.binary.version> in parent POM
+sed -i -e 's|<scala\.binary\.version>2.11<|<scala.binary.version>2.10<|1' pom.xml
```
Not sure, but can't this be done through profiles in maven ?

srowen replied:
The problem is how the POMs will be read by tools that just download the POMs to understand dependencies. They read the 'default' config. The alternative would be to edit the file to move `<activeByDefault>` to the other Scala profile, though that seemed harder.

ScrapCodes commented:
I am not a sed wizard, was just going through the man page. It does not say anything about -e being POSIX or non POSIX - sadly!. I do not have a mac, so no clues about the problem you stated.

srowen commented:
Don't worry about `-e`, I think. Does the command work to replace just the first occurrence on GNU `sed`? I can dig out a linux box but just wondered if you knew already. also, what's with `|` instead of `/`?

srowen committed:
Add note about non-GNU sed; fix new pom.xml update to work as intended on GNU sed

SparkQA commented:
[Test build #28261 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28261/consoleFull) for   PR 4876 at commit [`e875d4a`](https://github.com/apache/spark/commit/e875d4a4f896569fc894830b3bb782798212e7a9).
 * This patch merges cleanly.

srowen commented:
@ScrapCodes I updated the script to use more 'standard' `sed` syntax. The new command now works on GNU `sed`. Neither quite works on BSD `sed` (a la OS X) so I added a note about that, at least. I don't know how to fully fix that, and given this probably blocks 1.3, wanted to narrowly fix the immediate problem. WDYT?

srowen committed:
Oops, fixed reversed version numbers!

SparkQA commented:
[Test build #28262 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28262/consoleFull) for   PR 4876 at commit [`b060c44`](https://github.com/apache/spark/commit/b060c44604d0f383244f34307c79495bbf9269ab).
 * This patch merges cleanly.

ScrapCodes commented:
> what's with | instead of /?
sed let's you use any separator of your choice, I can not link you to documentation which validates this claim right away. I am not sure about its POSIX compliance either, (I can look it up in free time though.). For now, I don't see any harm in changing '|' '/' character.

ScrapCodes commented:
I did not run the script personally, but from the code it looks okay(as in harmless) to me. Sorry, I am not yet clear on how useful the addition of the last line is going to be. Because from your comment it is not clear 
1) what tools ?
2) How useful are those tools themselves ?
And how does the tool support block 1.3 release ?

srowen commented:
@ScrapCodes Right now, the published POMs are incorrect for Scala 2.11, because the parent POM defines `scala.binary.version` as 2.10, and child modules use that value to describe their dependencies. So the issue in this JIRA arises.

Of course, the POM has a profile to switch this value for Scala 2.11. But this has no effect on how anyone consuming these POMs parse the POMs. They just read the default value instead that we've published for Scala 2.11. That is, when any tool (Maven etc) reads these POMs, they do so as if without any special flags, vars, profiles activated.

So we also have to edit this value manually when making the 2.11 POM.

It's not a problem with tools; the dependencies described by the Spark Scala-2.11 POMs are wrong as a result of this, and any tool that reads them correctly reads the incorrect dependencies.

SparkQA commented:
[Test build #28261 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28261/consoleFull) for   PR 4876 at commit [`e875d4a`](https://github.com/apache/spark/commit/e875d4a4f896569fc894830b3bb782798212e7a9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28261/
Test PASSed.

SparkQA commented:
[Test build #28262 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28262/consoleFull) for   PR 4876 at commit [`b060c44`](https://github.com/apache/spark/commit/b060c44604d0f383244f34307c79495bbf9269ab).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28262/
Test PASSed.

ScrapCodes commented:
Hmm.. I was trusting our effective pom trick was working properly. Forgot
the JIRA id for that though.
On Mar 4, 2015 6:10 PM, "Sean Owen" <notifications@github.com> wrote:

> @ScrapCodes <https://github.com/ScrapCodes> Right now, the published POMs
> are incorrect for Scala 2.11, because the parent POM defines
> scala.binary.version as 2.10, and child modules use that value to
> describe their dependencies. So the issue in this JIRA arises.
>
> Of course, the POM has a profile to switch this value for Scala 2.11. But
> this has no effect on how anyone consuming these POMs parse the POMs. They
> just read the default value instead that we've published for Scala 2.11.
> That is, when any tool (Maven etc) reads these POMs, they do so as if
> without any special flags, vars, profiles activated.
>
> So we also have to edit this value manually when making the 2.11 POM.
>
> It's not a problem with tools; the dependencies described by the Spark
> Scala-2.11 POMs are wrong as a result of this, and any tool that reads them
> correctly reads the incorrect dependencies.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/4876#issuecomment-77150739>.
>

ScrapCodes commented:
Based on your comments and this issue I am going to take a deeper look tomorrow again. Meanwhile if you are rushed with respect to this PR. Looks like a good immediate solution.

pwendell commented:
I commented on the JIRA. This LGTM as an immediate fix - clearly the property is not correct in the published 2.11 poms.
